File: 2015/Day6/main.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <iterator>
#include <numeric>
#include <sstream>
#include <string>
#include <vector>

int process_light(const std::vector<std::string> &instructions) {
  std::vector<std::vector<bool>> grid(1000, std::vector<bool>(1000, false));

  for (const auto &instruction : instructions) {
    std::istringstream iss(instruction);
    std::vector<std::string> parts((std::istream_iterator<std::string>(iss)),
                                   std::istream_iterator<std::string>());

    if (parts[0] == "turn") {
      std::string action = parts[1];
      std::vector<int> start, end;
      std::stringstream ss1(parts[2]), ss2(parts[4]);
      std::string temp;
      while (std::getline(ss1, temp, ','))
        start.push_back(std::stoi(temp));
      while (std::getline(ss2, temp, ','))
        end.push_back(std::stoi(temp));

      for (int i = start[0]; i <= end[0]; ++i) {
        for (int j = start[1]; j <= end[1]; ++j) {
          grid[i][j] = (action == "on");
        }
      }
    } else if (parts[0] == "toggle") {
      std::vector<int> start, end;
      std::stringstream ss1(parts[1]), ss2(parts[3]);
      std::string temp;
      while (std::getline(ss1, temp, ','))
        start.push_back(std::stoi(temp));
      while (std::getline(ss2, temp, ','))
        end.push_back(std::stoi(temp));

      for (int i = start[0]; i <= end[0]; ++i) {
        for (int j = start[1]; j <= end[1]; ++j) {
          grid[i][j] = !grid[i][j];
        }
      }
    }
  }

  int count = 0;
  for (const auto &row : grid) {
    count +=
        std::count_if(row.begin(), row.end(), [](bool light) { return light; });
  }

  return count;
}

int process_light_brightness(const std::vector<std::string>& instructions) {
    std::vector<std::vector<int>> grid(1000, std::vector<int>(1000, 0));

    for (const auto& instruction : instructions) {
        std::istringstream iss(instruction);
        std::vector<std::string> parts((std::istream_iterator<std::string>(iss)),
                                       std::istream_iterator<std::string>());

        if (parts[0] == "turn") {
            std::string action = parts[1];
            std::vector<int> start, end;
            std::stringstream ss1(parts[2]), ss2(parts[4]);
            std::string temp;
            while (std::getline(ss1, temp, ','))
                start.push_back(std::stoi(temp));
            while (std::getline(ss2, temp, ','))
                end.push_back(std::stoi(temp));

            for (int i = start[0]; i <= end[0]; ++i) {
                for (int j = start[1]; j <= end[1]; ++j) {
                    if (action == "on") {
                        grid[i][j]++;
                    } else {
                        grid[i][j] = std::max(0, grid[i][j] - 1);
                    }
                }
            }
        } else if (parts[0] == "toggle") {
            std::vector<int> start, end;
            std::stringstream ss1(parts[1]), ss2(parts[3]);
            std::string temp;
            while (std::getline(ss1, temp, ','))
                start.push_back(std::stoi(temp));
            while (std::getline(ss2, temp, ','))
                end.push_back(std::stoi(temp));

            for (int i = start[0]; i <= end[0]; ++i) {
                for (int j = start[1]; j <= end[1]; ++j) {
                    grid[i][j] += 2;
                }
            }
        }
    }

    int total_brightness = 0;
    for (const auto& row : grid) {
        total_brightness += std::accumulate(row.begin(), row.end(), 0);
    }

    return total_brightness;
}

int main() {
    std::ifstream file("input.txt");
    std::vector<std::string> instructions;
    std::string line;

    while (std::getline(file, line)) {
        instructions.push_back(line);
    }

    std::cout << "Part 1: " << process_light(instructions) << std::endl;
    std::cout << "Part 2: " << process_light_brightness(instructions) << std::endl;

    return 0;
}

File: 2015/Day6/main.lua
local function process_lights(instructions)
    local grid = {}
    for i = 1, 1000 do
        grid[i] = {}
        for j = 1, 1000 do
            grid[i][j] = false
        end
    end

    for _, instruction in ipairs(instructions) do
        local parts = {}
        for part in instruction:gmatch("%S+") do
            table.insert(parts, part)
        end

        if parts[1] == "turn" then
            local action = parts[2]
            local start = { parts[3]:match("(%d+),(%d+)") }
            local end_ = { parts[5]:match("(%d+),(%d+)") }
            for i = tonumber(start[1]), tonumber(end_[1]) do
                for j = tonumber(start[2]), tonumber(end_[2]) do
                    grid[i + 1][j + 1] = action == "on"
                end
            end
        elseif parts[1] == "toggle" then
            local start = { parts[2]:match("(%d+),(%d+)") }
            local end_ = { parts[4]:match("(%d+),(%d+)") }
            for i = tonumber(start[1]), tonumber(end_[1]) do
                for j = tonumber(start[2]), tonumber(end_[2]) do
                    grid[i + 1][j + 1] = not grid[i + 1][j + 1]
                end
            end
        end
    end

    local count = 0
    for i = 1, 1000 do
        for j = 1, 1000 do
            if grid[i][j] then
                count = count + 1
            end
        end
    end

    return count
end

local function process_light_brightness(instructions)
    local grid = {}
    for i = 1, 1000 do
        grid[i] = {}
        for j = 1, 1000 do
            grid[i][j] = 0
        end
    end

    for _, instruction in ipairs(instructions) do
        local parts = {}
        for part in instruction:gmatch("%S+") do
            table.insert(parts, part)
        end

        if parts[1] == "turn" then
            local action = parts[2]
            local start = { parts[3]:match("(%d+),(%d+)") }
            local end_ = { parts[5]:match("(%d+),(%d+)") }
            for i = tonumber(start[1]), tonumber(end_[1]) do
                for j = tonumber(start[2]), tonumber(end_[2]) do
                    if action == "on" then
                        grid[i + 1][j + 1] = grid[i + 1][j + 1] + 1
                    elseif action == "off" then
                        grid[i + 1][j + 1] = math.max(0, grid[i + 1][j + 1] - 1)
                    end
                end
            end
        elseif parts[1] == "toggle" then
            local start = { parts[2]:match("(%d+),(%d+)") }
            local end_ = { parts[4]:match("(%d+),(%d+)") }
            for i = tonumber(start[1]), tonumber(end_[1]) do
                for j = tonumber(start[2]), tonumber(end_[2]) do
                    grid[i + 1][j + 1] = grid[i + 1][j + 1] + 2
                end
            end
        end
    end

    local total_brightness = 0
    for i = 1, 1000 do
        for j = 1, 1000 do
            total_brightness = total_brightness + grid[i][j]
        end
    end

    return total_brightness
end

local instructions = {}
for line in io.lines('input.txt') do
    table.insert(instructions, line)
end

print('Part 1: ' .. process_lights(instructions))
print('Part 2: ' .. process_light_brightness(instructions))
